app/services/import_notifier.rb

```diff
@@ -4,15 +4,41 @@ def initialize(description: nil)
   end
 
   def meters_running_behind
-    meters = Meter.active
-    .joins(:amr_data_feed_readings)
-    .where("amr_data_feed_readings.created_at >= NOW() - INTERVAL '1 year'")
-    .joins("INNER JOIN amr_data_feed_configs on amr_data_feed_readings.amr_data_feed_config_id = amr_data_feed_configs.id")
-    .where.not("amr_data_feed_configs.import_warning_days" => nil)
-    .joins(:amr_validated_readings)
-    .group('meters.id')
-    .having("MAX(amr_validated_readings.reading_date) < NOW() - MIN(amr_data_feed_configs.import_warning_days) * '1 day'::interval")
-    meters.sort_by {|m| [m.school.area_name, m.meter_type, m.school_name, m.mpan_mprn]}
+    find_meters_running_behind.sort_by do |meter|
+      [
+        meter&.school&.area_name || '',
+        meter&.meter_type || '',
+        meter&.school_name || '',
+        meter&.mpan_mprn || 0
+      ]
+    end
+  end
+
+  def find_meters_running_behind
```

Comment:
This doesn't look quite right to me.

Firstly, if we're finding the warning period using the data source (or a global default) then we don't need to join to `amr_data_feed_readings`. Avoiding that join is the main motivator for this optimisation. So that should be removed.

Secondly, do you need the `MIN` in the `having` clause? In the current version I use MIN because there might be multiple warning periods, so I use the smallest. But in this version there will either be the warning period for the meters data source OR the global default as COALESCE returns the first non-null value. 

Finally, as we're later sorting the results using a `sort_by`, we probably don't need the `.order`?

Reply:
:+1: I've pushed some updates to the query

Reply:
`sort_by` fails on nil and (at least in my tests) the `meter&.school&.area_name` could return nil.  As all other fields are required i've updated the area_name method to return an empty string if no school group is present